README
NAME
    MooseX::Has::Sugar - Sugar Syntax for moose 'has' fields

VERSION
    version 0.0403

SYNOPSIS
    Moose "has" syntax is generally fine, but sometimes one gets bothered
    with the constant typing of string quotes for things. MooseX::Types
    exists and in many ways reduces the need for constant string creation.

  Primary Benefits at a Glance
   Reduced Typing in "has" declarations.
    The constant need to type "=>" and '' is fine for one-off cases, but the
    instant you have more than about 4 attributes it starts to get annoying.

   More compact declarations.
    Reduces much of the redundant typing in most cases, which makes your
    life easier, and makes it take up less visual space, which makes it
    faster to read.

   No String Worries
    Strings are often problematic, due to whitespace etc. Noted that if you
    do happen to mess them up, Moose should at *least* warn you that you've
    done something daft. Using this alleviates that worry.

  Before this Module.
   Classical Moose
        has foo => (
                isa => 'Str',
                is  => 'ro',
                required => 1,
        );

        has bar => (
                isa => 'Str',
                is => 'rw'
                lazy_build => 1,
        );

   Lazy Evil way to do it:
    PLEASE DONT DO THIS

        has qw( foo isa Str is ro required 1 );
        has qw( bar isa Str is rw lazy_build 1 );

  With this module
    ( and with MooseX::Types )

        use MooseX::Types::Moose qw( Str );
        use MooseX::Has::Sugar;

        has foo => (
                isa => Str,
                ro,
                required,
        );
        has bar => (
                isa => Str,
                rw,
                lazy_build,
        );

    Or even

        use MooseX::Types::Moose qw( Str );
        use MooseX::Has::Sugar;

        has foo => ( isa => Str, ro,  required, );
        has bar => ( isa => Str, rw,  lazy_build, );

  Alternative Forms
   Basic "is" Expansion Only
    ( using MooseX::Has::Sugar::Minimal instead )

        use MooseX::Types::Moose qw( Str );
        use MooseX::Has::Sugar::Minimal;

        has foo => (
                isa => Str,
                is  => ro,
                required => 1,
        );
        has bar => (
                isa => Str,
                is => rw,
                lazy_build => 1,
        );

   Attribute Expansions with Basic Expansions
    ( Combining parts of this and MooseX::Has::Sugar::Minimal )

        use MooseX::Types::Moose qw( Str );
        use MooseX::Has::Sugar::Minimal;
        use MooseX::Has::Sugar qw( :attrs );

        has foo => (
                isa => Str,
                is  => ro,
                required,
        );
        has bar => (
                isa => Str,
                is => rw,
                lazy_build,
        );

EXPORT
    rw
    ro
    bare
    lazy
    lazy_build
    required
    coerce
    weak_ref
    auto_deref

EXPORT GROUPS
    :default
        Since 0.0300, this exports all our syntax, the same as ":attrs
        :isattrs". Primarily because I found you generally want all the
        sugar, not just part of it. This also gets rid of that nasty
        exclusion logic.

    :is DEPRECATED. See MooseX::Has::Sugar::Minimal for the same
        functionality

    :attrs
        This exports "lazy" , "lazy_build" and "required", "coerce",
        "weak_ref" and "auto_deref" as subs that assume positive.

            has foo => (
                    required,
                    isa => 'Str',
            );

    :isattrs
        This exports "ro", "rw" and "bare" as lists, so they behave as
        stand-alone attrs like "lazy" does.

            has foo => (
                    required,
                    isa => 'Str',
                    ro,
            );

        NOTE: This option is incompatible with MooseX::Has::Sugar::Minimal

    :allattrs
        This is a shorthand for qw( :isattrs :attrs )

FUNCTIONS
    These you probably don't care about, they're all managed by
    Sub::Exporter and its stuff anyway.

    rw  returns "('is','rw')"

    ro  returns "('is','ro')"

    bare
        returns "('is','bare')"

    lazy
        returns "('lazy',1)"

    required
        returns "('required',1)"

    lazy_build
        returns "('lazy_build',1)"

    coerce
        returns "('coerce',1)"

    weak_ref
        returns "('weak_ref',1)"

    auto_deref
        returns "('auto_deref',1)"

BUGS
    Please report any bugs or feature requests to "bug-moosex-has-extras at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Has-Sugar>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc MooseX::Has::Sugar

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Has-Sugar>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/MooseX-Has-Sugar>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/MooseX-Has-Sugar>

    *   Search CPAN

        <http://search.cpan.org/dist/MooseX-Has-Sugar/>

ACKNOWLEDGEMENTS
AUTHOR
      Kent Fredric <kentnl at cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2009 by Kent Fredric.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for MooseX-Has-Sugar

0.0403    2009-06-30 13:56:07 UTC
          Using Dist::Zilla's handy author-tests feature
          Revised Docmentation a little to be more correct

0.0402    2009-06-29 19:43:05 UTC
          Fixed missing META.yml in Dzil build

0.0401    2009-06-29 18:16:51 UTC
          Fixed Dep on Moose Test.
          Moved to Dist::Zilla.
          Loads of edits for change

0.0400    Sun Jun 28 12:53:52 NZST 2009
          Improved Test cases
          Improved meta dependency advertising
          added 'bare' keyword.

0.0300    Sat May 30 04:22:57 NZST 2009
          export group :is/-is moved to ::Minimal.
          MX::H::S::Minimal exports by default
          MX::H::Sugar exports all list-flavours by default.
          MX::H::Sugar croaks if group :is is requested.
          Test/Documentation updated.

0.0200    Sun May 17 09:38:31 NZST 2009
          Fixed META.yml
          Added weak_ref, coerce and auto_deref to -attrs
          Added collision detection to complain if you use it wrong.
          Removed Constant Folding based subs, too pesky at present.
          Added A bunch of tests.

0.0100    Fri May 15 21:18:30 NZST 2009
          First version.



t/04_values/EMixed.pm
package EMixed;
our $VERSION = '0.0403';


# $Id:$
use strict;
use warnings;
use Moose;
use MooseX::Has::Sugar::Minimal;
use MooseX::Has::Sugar qw( :attrs );
use namespace::autoclean;

sub generated {
    {
        isa => 'Str',
        is  => ro,
        required, lazy, lazy_build, coerce, weak_ref, auto_deref
    };
}

sub manual {
    {
        isa        => 'Str',
        is         => 'ro',
        required   => 1,
        lazy       => 1,
        lazy_build => 1,
        coerce     => 1,
        weak_ref   => 1,
        auto_deref => 1,
    };
}

__PACKAGE__->meta->make_immutable;
1;



t/05_is/TestPackage.pm
package TestPackage;
our $VERSION = '0.0403';


# $Id:$
use strict;
use warnings;
use Moose;
use namespace::autoclean;

use MooseX::Has::Sugar::Minimal;

has roattr => ( isa => 'Str', is => ro, required => 1, );

has rwattr => ( isa => 'Str', is => rw, required => 1, );

has bareattr => ( isa => 'Str', is => bare, required => 1, );

__PACKAGE__->meta->make_immutable;

1;



t/04_values/AMinimal.pm
package AMinimal;
our $VERSION = '0.0403';


# $Id:$
use strict;
use warnings;

use Moose;
use MooseX::Has::Sugar::Minimal;
use namespace::autoclean;

sub ro_generated { { isa => 'Str', is => ro, } }

sub ro_manual { { isa => 'Str', is => 'ro', } }

sub rw_generated { { isa => 'Str', is => rw, } }

sub rw_manual { { isa => 'Str', is => 'rw', } }

sub bare_generated { { isa => 'Str', is => bare, } }

sub bare_manual { { isa => 'Str', is => 'bare', } }

__PACKAGE__->meta->make_immutable;
1;



lib/MooseX/Has/Sugar/Minimal.pm
package MooseX::Has::Sugar::Minimal;
our $VERSION = '0.0403';


# ABSTRACT: Less Sugary Syntax for moose 'has' fields

use warnings;
use strict;

use Sub::Exporter ();

Sub::Exporter::setup_exporter(
    {
        exports => [ 'ro', 'rw', 'bare', ],
        groups  => {
            is      => [ 'ro', 'rw', 'bare', ],
            default => ['-is'],
        }
    }
);

sub bare() {
    return ('bare');
}

sub ro() {
    return ('ro');
}

sub rw() {
    return ('rw');
}

1;




=pod

=head1 NAME

MooseX::Has::Sugar::Minimal - Less Sugary Syntax for moose 'has' fields

=head1 VERSION

version 0.0403

=head1 SYNOPSIS

This is a legacy variant of L<MooseX::Has::Sugar> which only exports C<ro>
and C<rw> functions, the way L<MooseX::Has::Sugar> used to with C<:is>;

    use MooseX::Types::Moose qw( Str );
    use MooseX::Has::Sugar::Minimal;

    has foo => (
            isa => Str,
            is  => ro,
            required => 1,
    );
    has bar => (
            isa => Str,
            is => rw,
            lazy_build => 1,
    );

=head1 EXPORT

=over 4

=item rw

=item ro

=item bare

=back 

=head1 EXPORT GROUPS

=over 4

=item :default

Exports C<:is>

=item :is

Exports C<ro> and C<rw> and C<bare>

=back 

=head1 CONFLICTS

This module is not intended to be used in conjunction with
 L<MooseX::Has::Sugar>.

We both export C<ro> and C<rw> in different ways.

If you do however want to use them in conjuction, specific imports must
 be done on L<MooseX::Has::Sugar>'s side to stop it exporting different
 ro/rw. Any of the below should be fine.

    use MooseX::Has::Sugar::Minimal;
    use MooseX::Has::Sugar qw( :attrs );

    has foo =>( is => rw , lazy_build );

    use MooseX::Has::Sugar::Minimal;
    use MooseX::Has::Sugar qw( lazy_build );

    has foo =>( is => rw , lazy_build );

=head1 FUNCTIONS

These you probably don't care about, they're all managed by L<Sub::Exporter>
 and its stuff anyway.

=over 4

=item rw

returns C<('rw')>

=item ro

returns C<('ro')>

=item bare

returns C<('bare')>

=back 

=head1 BUGS

Please report any bugs or feature requests to
C<bug-moosex-has-extras at rt.cpan.org>
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Has-Sugar>.

I will be notified, and then you'll automatically be notified of progress
 on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooseX::Has::Sugar::Minimal

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Has-Sugar>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooseX-Has-Sugar>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MooseX-Has-Sugar>

=item * Search CPAN

L<http://search.cpan.org/dist/MooseX-Has-Sugar/>

=back 

=head1 ACKNOWLEDGEMENTS



=head1 AUTHOR

  Kent Fredric <kentnl at cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Kent Fredric.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut 



__END__



lib/MooseX/Has/Sugar.pm
package MooseX::Has::Sugar;
our $VERSION = '0.0403';


# ABSTRACT: Sugar Syntax for moose 'has' fields

use warnings;
use strict;

use Carp          ();
use Sub::Exporter ();

Sub::Exporter::setup_exporter(
    {
        as      => 'do_import',
        exports => [ 'ro', 'rw', 'required', 'lazy', 'lazy_build', 'coerce', 'weak_ref', 'auto_deref', 'bare', ],
        groups  => {
            isattrs => [ 'ro',       'rw',   'bare', ],
            attrs   => [ 'required', 'lazy', 'lazy_build', 'coerce', 'weak_ref', 'auto_deref', ],
            allattrs => [ '-attrs', '-isattrs' ],
            default  => [ '-attrs', '-isattrs' ],
        }
    }
);

sub import {
    for (@_) {
        if ( $_ =~ qr/^[:-]is$/ ) {
            Carp::croak( "Trivial ro/rw with :is dropped as of 0.0300.\n" . " See MooseX::Has::Sugar::Minimal for those. " );
        }
    }
    goto &MooseX::Has::Sugar::do_import;
}

sub bare() {
    return ( 'is', 'bare' );
}

sub ro() {
    return ( 'is', 'ro' );
}

sub rw() {
    return ( 'is', 'rw' );
}

sub required() {
    return ( 'required', 1 );
}

sub lazy() {
    return ( 'lazy', 1 );
}

sub lazy_build() {
    return ( 'lazy_build', 1 );
}

sub weak_ref() {
    return ( 'weak_ref', 1 );
}

sub coerce() {
    return ( 'coerce', 1 );
}

sub auto_deref() {
    return ( 'auto_deref', 1 );
}
1;


__END__

=pod

=head1 NAME

MooseX::Has::Sugar - Sugar Syntax for moose 'has' fields

=head1 VERSION

version 0.0403

=head1 SYNOPSIS

L<Moose> C<has> syntax is generally fine, but sometimes one gets bothered with
the constant typing of string quotes for things. L<MooseX::Types> exists and in
many ways reduces the need for constant string creation.

=head2 Primary Benefits at a Glance

=head3 Reduced Typing in C<has> declarations.

The constant need to type C<=E<gt>> and C<''> is fine for one-off cases, but
the instant you have more than about 4 attributes it starts to get annoying.

=head3 More compact declarations.

Reduces much of the redundant typing in most cases, which makes your life easier,
and makes it take up less visual space, which makes it faster to read.

=head3 No String Worries

Strings are often problematic, due to whitespace etc. Noted that if you do
happen to mess them up, Moose should at I<least> warn you that you've done
something daft. Using this alleviates that worry.

=head2 Before this Module.

=head3 Classical Moose

    has foo => (
            isa => 'Str',
            is  => 'ro',
            required => 1,
    );

    has bar => (
            isa => 'Str',
            is => 'rw'
            lazy_build => 1,
    );

=head3 Lazy Evil way to do it:

B<PLEASE DONT DO THIS>

    has qw( foo isa Str is ro required 1 );
    has qw( bar isa Str is rw lazy_build 1 );

=head2 With this module

( and with MooseX::Types )

    use MooseX::Types::Moose qw( Str );
    use MooseX::Has::Sugar;

    has foo => (
            isa => Str,
            ro,
            required,
    );
    has bar => (
            isa => Str,
            rw,
            lazy_build,
    );

Or even

    use MooseX::Types::Moose qw( Str );
    use MooseX::Has::Sugar;

    has foo => ( isa => Str, ro,  required, );
    has bar => ( isa => Str, rw,  lazy_build, );

=head2 Alternative Forms

=head3 Basic C<is> Expansion Only

( using L<MooseX::Has::Sugar::Minimal> instead )

    use MooseX::Types::Moose qw( Str );
    use MooseX::Has::Sugar::Minimal;

    has foo => (
            isa => Str,
            is  => ro,
            required => 1,
    );
    has bar => (
            isa => Str,
            is => rw,
            lazy_build => 1,
    );

=head3 Attribute Expansions with Basic Expansions

( Combining parts of this and L<MooseX::Has::Sugar::Minimal> )

    use MooseX::Types::Moose qw( Str );
    use MooseX::Has::Sugar::Minimal;
    use MooseX::Has::Sugar qw( :attrs );

    has foo => (
            isa => Str,
            is  => ro,
            required,
    );
    has bar => (
            isa => Str,
            is => rw,
            lazy_build,
    );

=head1 EXPORT

=over 4

=item rw

=item ro

=item bare

=item lazy

=item lazy_build

=item required

=item coerce

=item weak_ref

=item auto_deref

=back 

=head1 EXPORT GROUPS

=over 4

=item :default

Since 0.0300, this exports all our syntax, the same as C<:attrs :isattrs>.
Primarily because I found you generally want all the sugar, not just part of it.
This also gets rid of that nasty exclusion logic.

=item :is

B<DEPRECATED>. See L<MooseX::Has::Sugar::Minimal> for the same functionality

=item :attrs

This exports C<lazy> , C<lazy_build> and C<required>, C<coerce>, C<weak_ref>
and C<auto_deref> as subs that assume positive.

    has foo => (
            required,
            isa => 'Str',
    );

=item :isattrs

This exports C<ro>, C<rw> and C<bare> as lists, so they behave as stand-alone attrs like
C<lazy> does.

    has foo => (
            required,
            isa => 'Str',
            ro,
    );

B<NOTE: This option is incompatible with L<MooseX::Has::Sugar::Minimal>>

=item :allattrs

This is  a shorthand for  qw( :isattrs :attrs )

=back 

=head1 FUNCTIONS

These you probably don't care about, they're all managed by L<Sub::Exporter>
and its stuff anyway.

=over 4

=item rw

returns C<('is','rw')>

=item ro

returns C<('is','ro')>

=item bare

returns C<('is','bare')>

=item lazy

returns C<('lazy',1)>

=item required

returns C<('required',1)>

=item lazy_build

returns C<('lazy_build',1)>

=item coerce

returns C<('coerce',1)>

=item weak_ref

returns C<('weak_ref',1)>

=item auto_deref

returns C<('auto_deref',1)>

=back 

=head1 BUGS

Please report any bugs or feature requests to
C<bug-moosex-has-extras at rt.cpan.org>, or through
the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=MooseX-Has-Sugar>.  I will be
notified, and then you'll automatically be notified of progress on your bug
as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc MooseX::Has::Sugar

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=MooseX-Has-Sugar>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/MooseX-Has-Sugar>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/MooseX-Has-Sugar>

=item * Search CPAN

L<http://search.cpan.org/dist/MooseX-Has-Sugar/>

=back 

=head1 ACKNOWLEDGEMENTS



=head1 AUTHOR

  Kent Fredric <kentnl at cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2009 by Kent Fredric.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut 


